lib/instrumentation/index.ts

```diff
@@ -0,0 +1,147 @@
+      [SemanticResourceAttributes.SERVICE_VERSION]: pkg.version,
+    }),
+  });
+
+  // add processors
+  if (isTraceDebuggingEnabled()) {
+    traceProvider.addSpanProcessor(
+      new SimpleSpanProcessor(new ConsoleSpanExporter())
+    );
+  }
+
+  // OTEL specification environment variable
+  if (isTraceSendingEnabled()) {
+    const exporter = new OTLPTraceExporter();
+    traceProvider.addSpanProcessor(new BatchSpanProcessor(exporter));
+  }
+
+  const contextManager = new AsyncLocalStorageContextManager();
+  traceProvider.register({
+    contextManager,
+  });
+
+  registerInstrumentations({
+    instrumentations: [
+      new HttpInstrumentation({
+        applyCustomAttributesOnSpan: /* istanbul ignore next */ (
+          span,
+          request,
+          response
+        ) => {
+          // ignore 404 errors when the branch protection of Github could not be found. This is expected if no rules are configured
+          if (
+            !(request instanceof ClientRequest) ||
+            (request.host === `api.github.com` &&
+              request.path.endsWith(`/protection`) &&
+              response.statusCode === 404)
+          ) {
+            span.setStatus({ code: SpanStatusCode.OK });
+          }
+        },
+      }),
+      new BunyanInstrumentation(),
+    ],
+  });
+}
+
+/* istanbul ignore next */
+// https://github.com/open-telemetry/opentelemetry-js-api/issues/34
+export async function shutdown(): Promise<void> {
+  const traceProvider = getTracerProvider();
+  if (traceProvider instanceof NodeTracerProvider) {
+    await traceProvider.shutdown();
+  } else if (traceProvider instanceof ProxyTracerProvider) {
+    const delegateProvider = traceProvider.getDelegate();
+    if (delegateProvider instanceof NodeTracerProvider) {
+      await delegateProvider.shutdown();
+    }
+  }
+}
+
+export function getTracerProvider(): TracerProvider {
+  return api.trace.getTracerProvider();
+}
+
+function getTracer(): Tracer {
+  return getTracerProvider().getTracer('renovate');
+}
+
+export async function instrument<F extends () => Promise<unknown>>(
+  name: string,
+  fn: F
+): Promise<ReturnType<F>>;
+export async function instrument<F extends () => Promise<unknown>>(
+  name: string,
+  fn: F,
+  options: SpanOptions
+): Promise<ReturnType<F>>;
+export async function instrument<F extends () => Promise<unknown>>(
+  name: string,
+  fn: F,
+  options?: SpanOptions,
+  context?: Context
+): Promise<ReturnType<F>> {
+  const massagedOptions = options ?? {};
+  const massagedContext = context ?? api.context.active();
+
+  return await getTracer().startActiveSpan(
+    name,
+    massagedOptions,
+    massagedContext,
+    async (span) => {
+      try {
+        const result = await fn();
+        span.end();
+        return result;
+      } catch (e) {
+        span.setStatus({
+          code: SpanStatusCode.ERROR,
+          message: e,
+        });
+        span.end();
```

Comment:
maybe move  `span.end()` to finally block?

Reply:
From my understanding the finally block is not called if the error is rethrown in the catch block. I have preferred this instead of nested try and catch blocks.

Reply:
nope, finnaly block is always[^1] called

[^1]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Control_flow_and_error_handling#the_finally_block